### ISSUE TITLE
Failure to compile mercury with gcc 12

### ISSUE THREAD

**Issue by keriharris** [Mercury-Language/mercury#102]
The current stable+rotd releases of mercury (22.01.1 and rotd-2022-06-16) fail to compile with gcc 12.

e.g. on Ubuntu 22.04 LTS:

```
$ sudo apt-get install gcc-12
$ cd mercury-srcdist-rotd-2022-06-16
$ ./configure --with-cc=/usr/bin/gcc-12 --disable-most-grades --disable-debug-grades
...
gmake[2]: Entering directory '/home/keri/mercury/mercury-srcdist-rotd-2022-06-16/library'
../scripts/mgnuc --grade hlc.gc.pregen       --      -c array.c -o array.o
In function ‘mercury__array__lookup_3_p_0’,
    inlined from ‘mercury__array__copy_subarray_6_p_0’ at array.m:19641:7,
    inlined from ‘mercury__array__samsort_down_8_p_0’ at array.m:19405:9:
array.m:22271:27: error: array subscript -1 is below array bounds of ‘MR_Word[]’ {aka ‘long unsigned int[]’} [-Werror=array-bounds]
In file included from ../runtime/mercury.h:21,
                 from array.mih:31,
                 from array.c:29:
../runtime/mercury_types.h: In function ‘mercury__array__samsort_down_8_p_0’:
../runtime/mercury_types.h:128:29: note: while referencing ‘elements’
  128 |     MR_Word                 elements[MR_VARIABLE_SIZED];
      |                             ^~~~~~~~
In function ‘mercury__array__lookup_3_p_0’,
    inlined from ‘mercury__array__copy_subarray_6_p_0’ at array.m:19641:7,
    inlined from ‘mercury__array__samsort_up_8_p_0’ at array.m:19307:11:
array.m:22271:27: error: array subscript -1 is below array bounds of ‘MR_Word[]’ {aka ‘long unsigned int[]’} [-Werror=array-bounds]
../runtime/mercury_types.h: In function ‘mercury__array__samsort_up_8_p_0’:
../runtime/mercury_types.h:128:29: note: while referencing ‘elements’
  128 |     MR_Word                 elements[MR_VARIABLE_SIZED];
      |                             ^~~~~~~~
In function ‘mercury__array__lookup_3_p_0’,
    inlined from ‘mercury__array__f_84_121_112_101_83_112_101_99_79_102_95_95_112_114_101_100_95_95_99_111_112_121_95_115_117_98_97_114_114_97_121_95_95_91_84_32_61_32_105_110_116_93_95_48_95_49_6_p_0’ at array.m:3343:7,
    inlined from ‘mercury__array__f_84_121_112_101_83_112_101_99_79_102_95_95_112_114_101_100_95_95_115_97_109_115_111_114_116_95_100_111_119_110_95_95_91_84_32_61_32_105_110_116_93_95_48_95_49_8_p_0’ at array.m:2963:9:
array.m:22271:27: error: array subscript -1 is below array bounds of ‘MR_Word[]’ {aka ‘long unsigned int[]’} [-Werror=array-bounds]
../runtime/mercury_types.h: In function ‘mercury__array__f_84_121_112_101_83_112_101_99_79_102_95_95_112_114_101_100_95_95_115_97_109_115_111_114_116_95_100_111_119_110_95_95_91_84_32_61_32_105_110_116_93_95_48_95_49_8_p_0’:
../runtime/mercury_types.h:128:29: note: while referencing ‘elements’
  128 |     MR_Word                 elements[MR_VARIABLE_SIZED];
      |                             ^~~~~~~~
In function ‘mercury__array__lookup_3_p_0’,
    inlined from ‘mercury__array__f_84_121_112_101_83_112_101_99_79_102_95_95_112_114_101_100_95_95_99_111_112_121_95_115_117_98_97_114_114_97_121_95_95_91_84_32_61_32_105_110_116_93_95_48_95_49_6_p_0’ at array.m:3343:7,
    inlined from ‘mercury__array__f_84_121_112_101_83_112_101_99_79_102_95_95_112_114_101_100_95_95_115_97_109_115_111_114_116_95_117_112_95_95_91_84_32_61_32_105_110_116_93_95_48_95_49_8_p_0’ at array.m:2697:11:
array.m:22271:27: error: array subscript -1 is below array bounds of ‘MR_Word[]’ {aka ‘long unsigned int[]’} [-Werror=array-bounds]
../runtime/mercury_types.h: In function ‘mercury__array__f_84_121_112_101_83_112_101_99_79_102_95_95_112_114_101_100_95_95_115_97_109_115_111_114_116_95_117_112_95_95_91_84_32_61_32_105_110_116_93_95_48_95_49_8_p_0’:
../runtime/mercury_types.h:128:29: note: while referencing ‘elements’
  128 |     MR_Word                 elements[MR_VARIABLE_SIZED];
      |                             ^~~~~~~~
In function ‘mercury__array__lookup_3_p_0’,
    inlined from ‘mercury__array__f_84_121_112_101_83_112_101_99_79_102_95_95_112_114_101_100_95_95_99_111_112_121_95_115_117_98_97_114_114_97_121_95_95_91_84_32_61_32_115_116_114_105_110_103_93_95_48_95_49_6_p_0’ at array.m:3542:7,
    inlined from ‘mercury__array__f_84_121_112_101_83_112_101_99_79_102_95_95_112_114_101_100_95_95_115_97_109_115_111_114_116_95_100_111_119_110_95_95_91_84_32_61_32_115_116_114_105_110_103_93_95_48_95_49_8_p_0’ at array.m:3130:9:
array.m:22271:27: error: array subscript -1 is below array bounds of ‘MR_Word[]’ {aka ‘long unsigned int[]’} [-Werror=array-bounds]
../runtime/mercury_types.h: In function ‘mercury__array__f_84_121_112_101_83_112_101_99_79_102_95_95_112_114_101_100_95_95_115_97_109_115_111_114_116_95_100_111_119_110_95_95_91_84_32_61_32_115_116_114_105_110_103_93_95_48_95_49_8_p_0’:
../runtime/mercury_types.h:128:29: note: while referencing ‘elements’
  128 |     MR_Word                 elements[MR_VARIABLE_SIZED];
      |                             ^~~~~~~~
In function ‘mercury__array__lookup_3_p_0’,
    inlined from ‘mercury__array__f_84_121_112_101_83_112_101_99_79_102_95_95_112_114_101_100_95_95_99_111_112_121_95_115_117_98_97_114_114_97_121_95_95_91_84_32_61_32_115_116_114_105_110_103_93_95_48_95_49_6_p_0’ at array.m:3542:7,
    inlined from ‘mercury__array__f_84_121_112_101_83_112_101_99_79_102_95_95_112_114_101_100_95_95_115_97_109_115_111_114_116_95_117_112_95_95_91_84_32_61_32_115_116_114_105_110_103_93_95_48_95_49_8_p_0’ at array.m:2852:11:
array.m:22271:27: error: array subscript -1 is below array bounds of ‘MR_Word[]’ {aka ‘long unsigned int[]’} [-Werror=array-bounds]
../runtime/mercury_types.h: In function ‘mercury__array__f_84_121_112_101_83_112_101_99_79_102_95_95_112_114_101_100_95_95_115_97_109_115_111_114_116_95_117_112_95_95_91_84_32_61_32_115_116_114_105_110_103_93_95_48_95_49_8_p_0’:
../runtime/mercury_types.h:128:29: note: while referencing ‘elements’
  128 |     MR_Word                 elements[MR_VARIABLE_SIZED];
      |                             ^~~~~~~~
cc1: all warnings being treated as errors
gmake[2]: *** [/tmp/mmake.m8Dr8k:52239: array.o] Error 1
```

**Comment by juliensf** [comment]
It looks like the same workaround applied in commit d78fcae, also needs to be used with GCC 12.
(The whole thing is a bit stupid since it is basically GCC complaining about code that *it* has introduced.)

I'll commit a fix shortly.

**Comment by zsomogyi** [comment]
I would preemptively include gcc 1[0-9] in the test for the workaround, since
I wouldn't be surprised to find the same issue in gcc 13, 14 etc.

**Comment by juliensf** [comment]
@zsomogyi, shall do.

**Comment by juliensf** [comment]
Fixed in 22.01.2 and rotd-2022-06-19.
